
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>CanGraph.GraphifySMPDB.build_database &#8212; CanGraph 1.0 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script src="../../../_static/design-tabs.js"></script>
    <link rel="shortcut icon" href="../../../_static/cangraph-mini.png"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/cangraph-logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../index.html">
                    Main Page
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../Downloads.html">
   Downloads
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../Tutorials.html">
   Tutorials
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../Other%20Reading%20Material.html">
     Other Reading Material
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../Installing%20CanGraph.html">
     Installing CanGraph
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../Running%20the%20Software.html">
     Running the Software
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../Understanding%20its%20Outputs.html">
     Understanding its Outputs
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference external" href="https://www.pablomarcos.me/es/posts/master-en-biolog%c3%ada-computacional/tfm/">
   Presentations
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../CanGraph.html">
   CanGraph package
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../CanGraph.ExposomeExplorer.html">
     CanGraph.ExposomeExplorer package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../CanGraph.GraphifyDrugBank.html">
     CanGraph.GraphifyDrugBank package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../CanGraph.GraphifyHMDB.html">
     CanGraph.GraphifyHMDB package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../CanGraph.GraphifySMPDB.html">
     CanGraph.GraphifySMPDB package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../CanGraph.MeSHandMetaNetX.html">
     CanGraph.MeSHandMetaNetX package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../CanGraph.QueryWikidata.html">
     CanGraph.QueryWikidata package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../Install%20Apptainer.html">
     Install Apptainer
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../ToDoList.html">
   To-Do List
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Logo by <a href="https://www.instagram.com/danonino.caducado/"> Daniel Marcos </a> </br>
                     Website by <a href="https://www.pablomarcos.me/"> Pablo Marcos </a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/OMB-IARC/CanGraph/"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/OMB-IARC/CanGraph//issues/new?title=Issue%20on%20page%20%2F_modules/CanGraph/GraphifySMPDB/build_database.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1></h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <h1>Source code for CanGraph.GraphifySMPDB.build_database</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1"># SPDX-FileCopyrightText: 2022 Pablo Marcos &lt;software@loreak.org&gt;</span>
<span class="c1">#</span>
<span class="c1"># SPDX-License-Identifier: MIT</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">A python module that provides the necessary functions to transition the SMPDB database to graph format,</span>
<span class="sd">either from scratch importing all the nodes (as showcased in :obj:`CanGraph.GraphifySMPDB.main`) or in a case-by-case basis,</span>
<span class="sd">to annotate existing metabolites (as showcased in :obj:`CanGraph.main`).</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="c1"># Import external modules necessary for the script</span>
<span class="kn">import</span> <span class="nn">os</span><span class="o">,</span> <span class="nn">sys</span><span class="o">,</span> <span class="nn">shutil</span>               <span class="c1"># Vital modules to interact with the filesystem</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>                <span class="c1"># To parse FASTA files</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>                  <span class="c1"># Process tabular data</span>

<span class="c1"># Import subscripts for the program</span>
<span class="c1"># This hack that allows us to de-duplicate the miscleaneous script in this less-used script</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;../&quot;</span><span class="p">)</span>
<span class="c1"># .. NOTE:: Please beware that, if using this module by itself, you might need to copy &quot;miscelaneous.py&quot; into your path</span>
<span class="c1"># This is not the most elegant, but simplifies code maintenance, and this script shouldnt be used much so...</span>
<span class="kn">import</span> <span class="nn">miscelaneous</span> <span class="k">as</span> <span class="nn">misc</span>

<div class="viewcode-block" id="add_metabolites"><a class="viewcode-back" href="../../../CanGraph.GraphifySMPDB.html#CanGraph.GraphifySMPDB.build_database.add_metabolites">[docs]</a><span class="k">def</span> <span class="nf">add_metabolites</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adds &quot;Metabolite&quot; nodes to the database, according to individual CSVs present in the SMPDB website</span>

<span class="sd">    Args:</span>
<span class="sd">        tx          (neo4j.Session): The session under which the driver is running</span>
<span class="sd">        filename    (str): The name of the CSV file that is being imported</span>

<span class="sd">    Returns:</span>
<span class="sd">        neo4j.Result: A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</span>

<span class="sd">    .. NOTE:: Some of the node&#39;s properties might be set to &quot;null&quot; (important in order to work with it)</span>

<span class="sd">    .. NOTE:: This database clearly differentiates Metabolites and Proteins, so no overlap is accounted for</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        LOAD CSV WITH HEADERS FROM (&#39;file:///</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&#39;) AS line</span>
<span class="s2">        MERGE (m:Metabolite </span><span class="se">{{</span><span class="s2"> Metabolite_ID:line[&quot;Metabolite ID&quot;] </span><span class="se">}}</span><span class="s2">)</span>
<span class="s2">        SET m.Name = line[&quot;Metabolite Name&quot;], m.HMDB_ID = line[&quot;HMDB ID&quot;], m.KEGG_ID = line[&quot;KEGG ID&quot;], m.ChEBI_ID = line[&quot;ChEBI ID&quot;],</span>
<span class="s2">            m.DrugBank_ID = line[&quot;DrugBank ID&quot;], m.CAS_Number = line.CAS, m.Formula = line.Formula, m.IUPAC = line.IUPAC, m.SMILES = line.SMILES,</span>
<span class="s2">            m.InChI = line.InChI, m.InChIKey = line[&quot;InChI Key&quot;]</span>
<span class="s2">        MERGE (pa:Pathway </span><span class="se">{{</span><span class="s2"> SMPDB_ID:line[&quot;SMPDB ID&quot;] </span><span class="se">}}</span><span class="s2"> )</span>
<span class="s2">        SET pa.Name = line[&quot;Pathway Name&quot;], pa.Subject = line[&quot;Pathway Subject&quot;]</span>
<span class="s2">        MERGE (m)-[r:PART_OF_PATHWAY]-&gt;(pa)</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="add_proteins"><a class="viewcode-back" href="../../../CanGraph.GraphifySMPDB.html#CanGraph.GraphifySMPDB.build_database.add_proteins">[docs]</a><span class="k">def</span> <span class="nf">add_proteins</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adds &quot;Protein&quot; nodes to the database, according to individual CSVs present in the SMPDB website</span>

<span class="sd">    Args:</span>
<span class="sd">        tx          (neo4j.Session): The session under which the driver is running</span>
<span class="sd">        filename    (str): The name of the CSV file that is being imported</span>

<span class="sd">    Returns:</span>
<span class="sd">        neo4j.Result: A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</span>

<span class="sd">    .. NOTE:: Some of the node&#39;s properties might be set to &quot;null&quot; (important in order to work with it)</span>

<span class="sd">    .. NOTE:: This database clearly differentiates Metabolites and Proteins, so no overlap is accounted for</span>

<span class="sd">    .. TODO:: Why is the SMPDB_ID property called like that and not SMDB_ID?</span>

<span class="sd">    .. WARNING:: Since no unique identifier was found, CREATE had to be used (instead of merge). This might create</span>
<span class="sd">             duplicates. which should be accounted for.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        LOAD CSV WITH HEADERS FROM (&#39;file:///</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&#39;) AS line</span>
<span class="s2">        CREATE (p:Protein )</span>
<span class="s2">        SET p.Name = line[&quot;Protein Name&quot;], p.HMDB_ID = line[&quot;HMDBP ID&quot;], p.DrugBank_ID = line[&quot;DrugBank ID&quot;],</span>
<span class="s2">            p.Genbank_Protein_ID = line[&quot;GenBank ID&quot;], p.Gene_Name = line[&quot;Gene Name&quot;], p.Locus = line[&quot;Locus&quot;],</span>
<span class="s2">            p.UniProt_ID = line[&quot;Uniprot ID&quot;]</span>
<span class="s2">        MERGE (pa:Pathway </span><span class="se">{{</span><span class="s2"> SMPDB_ID:line[&quot;SMPDB ID&quot;] </span><span class="se">}}</span><span class="s2"> )</span>
<span class="s2">        SET pa.Name = line[&quot;Pathway Name&quot;], pa.Subject = line[&quot;Pathway Subject&quot;]</span>
<span class="s2">        MERGE (p)-[r:PART_OF_PATHWAY]-&gt;(pa)</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="add_pathways"><a class="viewcode-back" href="../../../CanGraph.GraphifySMPDB.html#CanGraph.GraphifySMPDB.build_database.add_pathways">[docs]</a><span class="k">def</span> <span class="nf">add_pathways</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adds &quot;Pathways&quot; nodes to the database, according to individual CSVs present in the SMPDB website</span>
<span class="sd">    Since this is done after the creation of said pathways in the last step, this will most likely just annotate them.</span>

<span class="sd">    Args:</span>
<span class="sd">        tx          (neo4j.Session): The session under which the driver is running</span>
<span class="sd">        filename    (str): The name of the CSV file that is being imported</span>

<span class="sd">    Returns:</span>
<span class="sd">        neo4j.Result: A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</span>

<span class="sd">    .. TODO:: This file is really big. It could be divided into smaller ones.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        LOAD CSV WITH HEADERS FROM (&#39;file:///</span><span class="si">{</span><span class="n">filename</span><span class="si">}</span><span class="s2">&#39;) AS line</span>
<span class="s2">        MERGE (pa:Pathway </span><span class="se">{{</span><span class="s2"> SMPDB_ID:line[&quot;SMPDB ID&quot;] </span><span class="se">}}</span><span class="s2"> )</span>
<span class="s2">        SET pa.Name = line[&quot;Name&quot;], pa.Category = line[&quot;Subject&quot;], pa.PW_ID = line[&quot;PW ID&quot;], pa.Description = line[&quot;Description&quot;]</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="add_sequence"><a class="viewcode-back" href="../../../CanGraph.GraphifySMPDB.html#CanGraph.GraphifySMPDB.build_database.add_sequence">[docs]</a><span class="k">def</span> <span class="nf">add_sequence</span><span class="p">(</span><span class="n">seq_id</span><span class="p">,</span> <span class="n">seq_name</span><span class="p">,</span> <span class="n">seq_type</span><span class="p">,</span> <span class="n">seq</span><span class="p">,</span> <span class="n">seq_format</span><span class="o">=</span><span class="s2">&quot;FASTA&quot;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adds &quot;Pathways&quot; nodes to the database, according to the sequences presented in FASTA files from the SMPDB website</span>

<span class="sd">    Args:</span>
<span class="sd">        tx          (neo4j.Session): The session under which the driver is running</span>
<span class="sd">        seq_id      (str): The UniProt Database Identifier for the sequence that is been imported</span>
<span class="sd">        seq_name    (str): The Name (i.e. FASTA header) of the Sequence that is been imported</span>
<span class="sd">        seq_type    (bool): The type of the sequence; can be either of [&quot;DNA&quot;, &quot;PROT&quot;]</span>
<span class="sd">        seq         (str): The seuqnce that is been imported; a text chain of nucleotides or aminoacids, identified by their acronyms</span>
<span class="sd">        seq_format  (str): The format the sequence is provided under; default is &quot;FASTA&quot;, but its optional</span>

<span class="sd">    Returns:</span>
<span class="sd">        neo4j.Result: A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        MERGE (s:Sequence </span><span class="se">{{</span><span class="s2"> UniProt_ID:&quot;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">&quot;, Name:&quot;</span><span class="si">{</span><span class="n">seq_name</span><span class="si">}</span><span class="s2">&quot;, Type:&quot;</span><span class="si">{</span><span class="n">seq_type</span><span class="si">}</span><span class="s2">&quot;, Sequence:&quot;</span><span class="si">{</span><span class="n">seq</span><span class="si">}</span><span class="s2">&quot;, Format:&quot;</span><span class="si">{</span><span class="n">seq_format</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">}}</span><span class="s2"> )</span>
<span class="s2">        MERGE (p:Protein </span><span class="se">{{</span><span class="s2"> UniProt_ID:&quot;</span><span class="si">{</span><span class="n">seq_id</span><span class="si">}</span><span class="s2">&quot; </span><span class="se">}}</span><span class="s2"> )</span>
<span class="s2">        MERGE (p)-[r:SEQUENCED_AS]-&gt;(s)</span>
<span class="s2">        &quot;&quot;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="build_from_file"><a class="viewcode-back" href="../../../CanGraph.GraphifySMPDB.html#CanGraph.GraphifySMPDB.build_database.build_from_file">[docs]</a><span class="k">def</span> <span class="nf">build_from_file</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">Neo4JImportPath</span><span class="p">,</span> <span class="n">driver</span><span class="p">,</span> <span class="n">filetype</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A function able to build a portion of the SMPDB in graph format, provided that one CSV is supplied to it.</span>
<span class="sd">    This CSVs are downloaded from the website, and can be presented either as the full file, or as a splitted</span>
<span class="sd">    version of it, with just one item per file (which is recommended due to memory limitations)</span>

<span class="sd">    Since file title represents a different pathway, the function automatically picks up and import the relative pathway node.</span>

<span class="sd">    Args:</span>
<span class="sd">        filepath (str): The path to the current file being imported</span>
<span class="sd">        Neo4JImportPath (str): The path from which Neo4J is importing data</span>
<span class="sd">        driver (neo4j.Driver): Neo4J&#39;s Bolt Driver currently in use</span>
<span class="sd">        filetype (bool): The type of file being imported; one of ether [&quot;Metabolite&quot;, &quot;Protein&quot;]- If the file is a FASTA sequence store, this will be auto-detected.</span>

<span class="sd">    Returns:</span>
<span class="sd">        This function modifies the Neo4J Database as desired, but does not produce any particular return.</span>

<span class="sd">     .. NOTE:: Since this adds a ton of low-resolution nodes, maybe have this db run first?</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span>
    <span class="n">databasepath</span> <span class="o">=</span> <span class="n">filepath</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/SMPDB/&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">pathway_id</span> <span class="o">=</span> <span class="n">filepath</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">all_the_proteins</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">databasepath</span><span class="p">)</span><span class="si">}</span><span class="s2">/SMPDB/smpdb_proteins/</span><span class="si">{</span><span class="n">pathway_id</span><span class="si">}</span><span class="s2">_proteins.csv&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">filetype</span> <span class="o">==</span> <span class="s2">&quot;Metabolite&quot;</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
            <span class="n">misc</span><span class="o">.</span><span class="n">manage_transaction</span><span class="p">(</span><span class="n">add_metabolites</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">driver</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">databasepath</span><span class="p">)</span><span class="si">}</span><span class="s2">/SMPDB/smpdb_proteins/</span><span class="si">{</span><span class="n">pathway_id</span><span class="si">}</span><span class="s2">_proteins.csv&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/corresponding.csv&quot;</span><span class="p">)</span>
            <span class="n">misc</span><span class="o">.</span><span class="n">manage_transaction</span><span class="p">(</span><span class="n">add_proteins</span><span class="p">(</span><span class="s2">&quot;corresponding.csv&quot;</span><span class="p">),</span> <span class="n">driver</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">filetype</span> <span class="o">==</span> <span class="s2">&quot;Protein&quot;</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
            <span class="n">misc</span><span class="o">.</span><span class="n">manage_transaction</span><span class="p">(</span><span class="n">add_proteins</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">driver</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">databasepath</span><span class="p">)</span><span class="si">}</span><span class="s2">/SMPDB/smpdb_metabolites/</span><span class="si">{</span><span class="n">pathway_id</span><span class="si">}</span><span class="s2">_metabolites.csv&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/corresponding.csv&quot;</span><span class="p">)</span>
            <span class="n">misc</span><span class="o">.</span><span class="n">manage_transaction</span><span class="p">(</span><span class="n">add_metabolites</span><span class="p">(</span><span class="s2">&quot;corresponding.csv&quot;</span><span class="p">),</span> <span class="n">driver</span><span class="p">)</span>

    <span class="c1"># We then import pathway info</span>
    <span class="n">all_pathways</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">databasepath</span><span class="p">)</span><span class="si">}</span><span class="s2">/SMPDB/smpdb_pathways.csv&quot;</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">just_this_pathway</span> <span class="o">=</span> <span class="n">all_pathways</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">all_pathways</span><span class="p">[</span><span class="s2">&quot;SMPDB ID&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pathway_id</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
    <span class="n">just_this_pathway</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/just_this_pathway.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


    <span class="c1"># And, finally, we add the protein sequences (They only reference proteins, not metabolites!):</span>
    <span class="n">genomic_sequences</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">databasepath</span><span class="p">)</span><span class="si">}</span><span class="s2">/SMPDB/smpdb_gene.fasta&quot;</span><span class="p">),</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">fasta</span> <span class="ow">in</span> <span class="n">genomic_sequences</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;)&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">all_the_proteins</span><span class="p">[</span><span class="s2">&quot;Uniprot ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">UniProt_ID</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;)&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">Name</span> <span class="o">=</span>  <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:])</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="o">+</span><span class="n">UniProt_ID</span><span class="o">+</span><span class="s2">&quot;)&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">Sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">seq</span><span class="p">);</span> <span class="n">Format</span> <span class="o">=</span> <span class="s2">&quot;FASTA&quot;</span><span class="p">;</span> <span class="n">Type</span> <span class="o">=</span> <span class="s2">&quot;DNA&quot;</span>
            <span class="k">with</span> <span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
                <span class="n">misc</span><span class="o">.</span><span class="n">manage_transaction</span><span class="p">(</span><span class="n">add_sequence</span><span class="p">(</span><span class="n">UniProt_ID</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">Type</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Format</span><span class="p">),</span> <span class="n">driver</span><span class="p">)</span>

    <span class="n">proteic_sequences</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">databasepath</span><span class="p">)</span><span class="si">}</span><span class="s2">/SMPDB/smpdb_protein.fasta&quot;</span><span class="p">),</span><span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">fasta</span> <span class="ow">in</span> <span class="n">proteic_sequences</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;)&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">all_the_proteins</span><span class="p">[</span><span class="s2">&quot;Uniprot ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">UniProt_ID</span> <span class="o">=</span> <span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;)&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">Name</span> <span class="o">=</span>  <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">:])</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="o">+</span><span class="n">UniProt_ID</span><span class="o">+</span><span class="s2">&quot;)&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">Sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">description</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">seq</span><span class="p">);</span> <span class="n">Format</span> <span class="o">=</span> <span class="s2">&quot;FASTA&quot;</span><span class="p">;</span> <span class="n">Type</span> <span class="o">=</span> <span class="s2">&quot;PROT&quot;</span>
            <span class="k">with</span> <span class="n">driver</span><span class="o">.</span><span class="n">session</span><span class="p">()</span> <span class="k">as</span> <span class="n">session</span><span class="p">:</span>
                <span class="n">misc</span><span class="o">.</span><span class="n">manage_transaction</span><span class="p">(</span><span class="n">add_sequence</span><span class="p">(</span><span class="n">UniProt_ID</span><span class="p">,</span> <span class="n">Name</span><span class="p">,</span> <span class="n">Type</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Format</span><span class="p">),</span> <span class="n">driver</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/just_this_pathway.csv&quot;</span><span class="p">);</span> <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/corresponding.csv&quot;</span><span class="p">);</span> <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">Neo4JImportPath</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>
</pre></div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Pablo Marcos<br/>
  
      &copy; Copyright 2022, Pablo Marcos &lt;software@loreak.org&gt;.<br/>
    Last updated on Dec 17, 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>