
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>CanGraph.GraphifySMPDB package &#8212; CanGraph 0.9 documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script src="_static/design-tabs.js"></script>
    <link rel="shortcut icon" href="_static/cangraph-mini.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="CanGraph.MeSHandMetaNetX package" href="CanGraph.MeSHandMetaNetX.html" />
    <link rel="prev" title="CanGraph.GraphifyHMDB package" href="CanGraph.GraphifyHMDB.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/cangraph-logo.png" class="logo" alt="logo">
      
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    Main Page
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="ToDoList.html">
   To-Do List
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="CanGraph.html">
   CanGraph package
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="CanGraph.ExposomeExplorer.html">
     CanGraph.ExposomeExplorer package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="CanGraph.GraphifyDrugBank.html">
     CanGraph.GraphifyDrugBank package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="CanGraph.GraphifyHMDB.html">
     CanGraph.GraphifyHMDB package
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     CanGraph.GraphifySMPDB package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="CanGraph.MeSHandMetaNetX.html">
     CanGraph.MeSHandMetaNetX package
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="CanGraph.QueryWikidata.html">
     CanGraph.QueryWikidata package
    </a>
   </li>
  </ul>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Logo by <a href="https://www.instagram.com/danonino.caducado/"> Daniel Marcos </a> </br>
                     Website by <a href="https://www.pablomarcos.me/"> Pablo Marcos </a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/OMB-IARC/CanGraph/"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/OMB-IARC/CanGraph//issues/new?title=Issue%20on%20page%20%2FCanGraph.GraphifySMPDB.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/OMB-IARC/CanGraph/edit/main/docs/CanGraph.GraphifySMPDB.rst"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/CanGraph.GraphifySMPDB.rst.txt"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.rst</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> In this page
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#important-notices">
   Important Notices
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-CanGraph.GraphifySMPDB.build_database">
   CanGraph.GraphifySMPDB.build_database module
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-CanGraph.GraphifySMPDB.main">
   CanGraph.GraphifySMPDB.main module
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>CanGraph.GraphifySMPDB package</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> In this page </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#important-notices">
   Important Notices
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-CanGraph.GraphifySMPDB.build_database">
   CanGraph.GraphifySMPDB.build_database module
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-CanGraph.GraphifySMPDB.main">
   CanGraph.GraphifySMPDB.main module
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="cangraph-graphifysmpdb-package">
<h1>CanGraph.GraphifySMPDB package<a class="headerlink" href="#cangraph-graphifysmpdb-package" title="Permalink to this headline">#</a></h1>
<!--
SPDX-FileCopyrightText: 2022 Pablo Marcos <software@loreak.org>

SPDX-License-Identifier: MIT
-->
<div align="center"> <img src="_static/smpdb_schema.png" width="50%"> </div>
<br>
<p>This script, created as part of my Master’s Intenship at IARC, transitions the <a class="reference external" href="http://smpdb.ca/">Small Molecule Pathway Database</a> (a high quality database containing associations betweem metabolites, proteins and metabolomic pathways) to Neo4J format in an automated way, providing an export in GraphML format.</p>
<p>To run, it uses <code class="docutils literal notranslate"><span class="pre">alive_progress</span></code> to generate an interactive progress bar (that shows the script is still running through its most time-consuming parts) and the <code class="docutils literal notranslate"><span class="pre">neo4j</span></code> python driver. This requirements can be installed using: <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">-r</span> <span class="pre">requirements.txt</span></code>.</p>
<p>To run the script itself, use:</p>
<p><code class="docutils literal notranslate"><span class="pre">python3</span> <span class="pre">main.py</span> <span class="pre">neo4jadress</span> <span class="pre">databasename</span> <span class="pre">databasepassword</span></code></p>
<p>where:</p>
<ul class="simple">
<li><p><strong>neo4jadress</strong>: is the URL of the database, in neo4j:// or bolt:// format</p></li>
<li><p><strong>databasename</strong>: the name of the database in use. If using the free version, there will only be one database per project (neo4j being the default name); if using the pro version, you can specify an alternate name here</p></li>
<li><p><strong>databasepassword</strong>: the passowrd for the <strong>databasename</strong> DataBase. Since the arguments are passed by BaSH onto python3, you might need to escape special characters</p></li>
</ul>
<p>NOTE: The files will be downloaded to ./csvfolder, so please run the script somewhere you have read/write permissions</p>
<p>An archived version of this repository that takes into account the gitignored files can be created using: <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">archive</span> <span class="pre">HEAD</span> <span class="pre">-o</span> <span class="pre">${PWD##*/}.zip</span></code></p>
<section id="important-notices">
<h2>Important Notices<a class="headerlink" href="#important-notices" title="Permalink to this headline">#</a></h2>
<ul class="simple">
<li><p>Please ensure you have internet access, enough espace in your hard drive (around 5 GB) and read-write access in <code class="docutils literal notranslate"><span class="pre">./csvfolder</span></code>. The files needed to build the database will be stored there.</p></li>
<li><p>Since the “Structure” files at smpdb.ca seemed to be super complicated to import, we decided against doing so. However, regarding the future, they shouldn’t be overlooked, as they might include useful info</p></li>
<li><p>The “PW ID” column from the “Pathways” table, and the “Pathway Subject” from the “Metabolite” tables may correlate (both start with PW). However, they seem to use different formats, so we have decided against including them in the Final Database</p></li>
</ul>
</section>
<hr class="docutils" />
<p>The package consists of the following modules:</p>
<section id="module-CanGraph.GraphifySMPDB.build_database">
<span id="cangraph-graphifysmpdb-build-database-module"></span><h2>CanGraph.GraphifySMPDB.build_database module<a class="headerlink" href="#module-CanGraph.GraphifySMPDB.build_database" title="Permalink to this headline">#</a></h2>
<p>A python module that provides the necessary functions to transition the SMPDB database to graph format,
either from scratch importing all the nodes (as showcased in <a class="reference internal" href="#module-CanGraph.GraphifySMPDB.main" title="CanGraph.GraphifySMPDB.main"><code class="xref py py-obj docutils literal notranslate"><span class="pre">CanGraph.GraphifySMPDB.main</span></code></a>) or in a case-by-case basis,
to annotate existing metabolites (as showcased in <a class="reference internal" href="CanGraph.html#module-CanGraph.main" title="CanGraph.main"><code class="xref py py-obj docutils literal notranslate"><span class="pre">CanGraph.main</span></code></a>).</p>
<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.build_database.add_metabolites">
<span class="sig-name descname"><span class="pre">add_metabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.build_database.add_metabolites" title="Permalink to this definition">#</a></dt>
<dd><p>Adds “Metabolite” nodes to the database, according to individual CSVs present in the SMPDB website</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tx</strong> (<a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Session" title="(in Neo4j Python Driver v4.4)"><em>neo4j.work.simple.Session</em></a>) – The session under which the driver is running</p></li>
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The name of the CSV file that is being imported</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Result" title="(in Neo4j Python Driver v4.4)">neo4j.work.result.Result</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Some of the node’s properties might be set to “null” (important in order to work with it)</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This database clearly differentiates Metabolites and Proteins, so no overlap is accounted for</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.build_database.add_pathways">
<span class="sig-name descname"><span class="pre">add_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.build_database.add_pathways" title="Permalink to this definition">#</a></dt>
<dd><p>Adds “Pathways” nodes to the database, according to individual CSVs present in the SMPDB website
Since this is done after the creation of said pathways in the last step, this will most likely just annotate them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tx</strong> (<a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Session" title="(in Neo4j Python Driver v4.4)"><em>neo4j.work.simple.Session</em></a>) – The session under which the driver is running</p></li>
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The name of the CSV file that is being imported</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Result" title="(in Neo4j Python Driver v4.4)">neo4j.work.result.Result</a></p>
</dd>
</dl>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This file is really big. It could be divided into smaller ones.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.build_database.add_proteins">
<span class="sig-name descname"><span class="pre">add_proteins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.build_database.add_proteins" title="Permalink to this definition">#</a></dt>
<dd><p>Adds “Protein” nodes to the database, according to individual CSVs present in the SMPDB website</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tx</strong> (<a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Session" title="(in Neo4j Python Driver v4.4)"><em>neo4j.work.simple.Session</em></a>) – The session under which the driver is running</p></li>
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The name of the CSV file that is being imported</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Result" title="(in Neo4j Python Driver v4.4)">neo4j.work.result.Result</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Some of the node’s properties might be set to “null” (important in order to work with it)</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This database clearly differentiates Metabolites and Proteins, so no overlap is accounted for</p>
</div>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>Why is the SMPDB_ID property called like that and not SMDB_ID?</p>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Since no unique identifier was found, CREATE had to be used (instead of merge). This might create
duplicates. which should be accounted for.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.build_database.add_sequence">
<span class="sig-name descname"><span class="pre">add_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FASTA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.build_database.add_sequence" title="Permalink to this definition">#</a></dt>
<dd><p>Adds “Pathways” nodes to the database, according to the sequences presented in FASTA files from the SMPDB website</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tx</strong> (<a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Session" title="(in Neo4j Python Driver v4.4)"><em>neo4j.work.simple.Session</em></a>) – The session under which the driver is running</p></li>
<li><p><strong>seq_id</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The UniProt Database Identifier for the sequence that is been imported</p></li>
<li><p><strong>seq_name</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The Name (i.e. FASTA header) of the Sequence that is been imported</p></li>
<li><p><strong>seq_type</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – The type of the sequence; can be either of [“DNA”, “PROT”]</p></li>
<li><p><strong>seq</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The seuqnce that is been imported; a text chain of nucleotides or aminoacids, identified by their acronyms</p></li>
<li><p><strong>seq_format</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The format the sequence is provided under; default is “FASTA”, but its optional</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Neo4J connexion to the database that modifies it according to the CYPHER statement contained in the function.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Result" title="(in Neo4j Python Driver v4.4)">neo4j.work.result.Result</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.build_database.build_from_file">
<span class="sig-name descname"><span class="pre">build_from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">databasepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Neo4JImportPath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">driver</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filetype</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.build_database.build_from_file" title="Permalink to this definition">#</a></dt>
<dd><p>A function able to build a portion of the SMPDB in graph format, provided that one CSV is supplied to it.
This CSVs are downloaded from the website, and can be presented either as the full file, or as a splitted
version of it, with just one item per file (which is recommended due to memory limitations)</p>
<p>Since file title represents a different pathway, the function automatically picks up and import the relative pathway node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>databasepath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path to the database where all SMPDB CSVs are stored</p></li>
<li><p><strong>filepath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path to the current file being imported</p></li>
<li><p><strong>Neo4JImportPath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path from which Neo4J is importing data</p></li>
<li><p><strong>driver</strong> (<a class="reference external" href="https://neo4j.com/docs/api/python-driver/current/api.html#neo4j.Driver" title="(in Neo4j Python Driver v4.4)"><em>neo4j.Driver</em></a>) – Neo4J’s Bolt Driver currently in use</p></li>
<li><p><strong>filetype</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/functions.html#bool" title="(in Python v3.8)"><em>bool</em></a>) – The type of file being imported; one of ether [“Metabolite”, “Protein”]- If the file is a FASTA sequence store, this will be auto-detected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>This function modifies the Neo4J Database as desired, but does not produce any particular return.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Since this adds a ton of low-resolution nodes, maybe have this db run first?</p>
</div>
</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-CanGraph.GraphifySMPDB.main">
<span id="cangraph-graphifysmpdb-main-module"></span><h2>CanGraph.GraphifySMPDB.main module<a class="headerlink" href="#module-CanGraph.GraphifySMPDB.main" title="Permalink to this headline">#</a></h2>
<p>A python module that leverages the functions present in the <a class="reference internal" href="#module-CanGraph.GraphifySMPDB.build_database" title="CanGraph.GraphifySMPDB.build_database"><code class="xref py py-obj docutils literal notranslate"><span class="pre">build_database</span></code></a>
module to recreate <a class="reference external" href="http://smpdb.ca/">the SMPDB database</a> using a graph format and Neo4J,
and then provides an GraphML export file.</p>
<p>Please note that, to work, the functions here pre-suppose you have internet access, which will be used to download
HMDB’s CSVs under <code class="docutils literal notranslate"><span class="pre">`./csvfolder/`</span></code> (please ensure you have read-write access there).</p>
<p>For more details on how to run this script, please consult the package’s README</p>
<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.main.import_genomic_seqs">
<span class="sig-name descname"><span class="pre">import_genomic_seqs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.main.import_genomic_seqs" title="Permalink to this definition">#</a></dt>
<dd><p>Imports the <code class="docutils literal notranslate"><span class="pre">`smpdb_gene.fasta`</span></code> file from the SMPDB Database.
The function assumes the file is availaible at <code class="docutils literal notranslate"><span class="pre">`./csvfolder/smpdb_gene.fasta`</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>Neo4JImportPath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path from which Neo4J is importing data</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>This function modifies the Neo4J Database as desired, but does not produce any particular return.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.main.import_metabolites">
<span class="sig-name descname"><span class="pre">import_metabolites</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Neo4JImportPath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.main.import_metabolites" title="Permalink to this definition">#</a></dt>
<dd><p>Imports “Metabolite” files from the SMPDB Database.
The function assumes <code class="docutils literal notranslate"><span class="pre">`filename`</span></code> is availaible at <code class="docutils literal notranslate"><span class="pre">`./csvfolder/smpdb_metabolites/`</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The name of the CSV file that is being imported</p></li>
<li><p><strong>Neo4JImportPath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path from which Neo4J is importing data</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>This function modifies the Neo4J Database as desired, but does not produce any particular return.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.main.import_pathways">
<span class="sig-name descname"><span class="pre">import_pathways</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Neo4JImportPath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.main.import_pathways" title="Permalink to this definition">#</a></dt>
<dd><p>Imports the <code class="docutils literal notranslate"><span class="pre">`smpdb_pathways.csv`</span></code> file from the SMPDB Database.
The function assumes the file is availaible at <code class="docutils literal notranslate"><span class="pre">`./csvfolder/smpdb_gene.fasta`</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>Neo4JImportPath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path from which Neo4J is importing data</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>This function modifies the Neo4J Database as desired, but does not produce any particular return.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.main.import_proteic_seqs">
<span class="sig-name descname"><span class="pre">import_proteic_seqs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.main.import_proteic_seqs" title="Permalink to this definition">#</a></dt>
<dd><p>Imports the <code class="docutils literal notranslate"><span class="pre">`smpdb_protein.fasta`</span></code> file from the SMPDB Database.
The function assumes the file is availaible at <code class="docutils literal notranslate"><span class="pre">`./csvfolder/smpdb_protein.fasta`</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>Neo4JImportPath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path from which Neo4J is importing data</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>This function modifies the Neo4J Database as desired, but does not produce any particular return.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.main.import_proteins">
<span class="sig-name descname"><span class="pre">import_proteins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Neo4JImportPath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.main.import_proteins" title="Permalink to this definition">#</a></dt>
<dd><p>Imports “Protein” files from the SMPDB Database.
The function assumes <code class="docutils literal notranslate"><span class="pre">`filename`</span></code> is availaible at <code class="docutils literal notranslate"><span class="pre">`./csvfolder/smpdb_proteins/`</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The name of the CSV file that is being imported</p></li>
<li><p><strong>Neo4JImportPath</strong> (<a class="reference external" href="https://docs.python.org/3.8/library/stdtypes.html#str" title="(in Python v3.8)"><em>str</em></a>) – The path from which Neo4J is importing data</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>This function modifies the Neo4J Database as desired, but does not produce any particular return.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="CanGraph.GraphifySMPDB.main.main">
<span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#CanGraph.GraphifySMPDB.main.main" title="Permalink to this definition">#</a></dt>
<dd><p>The function that executes the code</p>
</dd></dl>

</section>
</section>


              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="CanGraph.GraphifyHMDB.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">CanGraph.GraphifyHMDB package</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="CanGraph.MeSHandMetaNetX.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">CanGraph.MeSHandMetaNetX package</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Pablo Marcos<br/>
  
      &copy; Copyright 2022, Pablo Marcos &lt;software @ loreak . org&gt;.<br/>
    Last updated on Aug 30, 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>